t/08-l1_serializer.t

```diff
@@ -0,0 +1,417 @@
+GET /t
+--- response_body_like
+l1_serializer threw an error: .*?: cannot transform
+--- no_error_log
+[error]
+
+
+
+=== TEST 5: l1_serializer is called in protected mode (L2 hit)
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local mlcache = require "resty.mlcache"
+
+            local called = false
+            local cache, err = mlcache.new("my_mlcache", "cache_shm", {
+                l1_serializer = function(s)
+                    if called then error("cannot transform") end
+                    called = true
+                    return string.format("transform(%q)", s)
+                end,
+            })
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            assert(cache:get("key", nil, function() return "foo" end))
+            cache.lru:delete("key")
+
+            local data, err = cache:get("key", nil, function() return "foo" end)
+            if not data then
+                ngx.say(err)
+            end
+            ngx.say(data)
+        }
+    }
+--- request
+GET /t
+--- response_body_like
+l1_serializer threw an error: .*?: cannot transform
+--- no_error_log
+[error]
+
+
+
+=== TEST 6: l1_serializer is not supposed to return a nil value
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local mlcache = require "resty.mlcache"
+
+            local cache, err = mlcache.new("my_mlcache", "cache_shm", {
+                l1_serializer = function(s)
+                    return nil
+                end,
+            })
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            local ok, err = cache:get("key", nil, function() return "foo" end)
+            assert(not ok, "get call returned successfully")
+            ngx.say(err)
+        }
+    }
+--- request
+GET /t
+--- response_body_like
+l1_serializer returned a nil value
+--- no_error_log
+[error]
+
+
+
+=== TEST 7: l1_serializer can return an error
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua_block {
+            local mlcache = require "resty.mlcache"
+
+            local cache, err = mlcache.new("my_mlcache", "cache_shm", {
+                l1_serializer = function(s)
+                    return nil, "cannot transform"
+                end,
+            })
+            if not cache then
+                ngx.log(ngx.ERR, err)
+                return
+            end
+
+            local data, err = cache:get("key", nil, function() return "foo" end)
+            if not data then
+                ngx.say(err)
+            end
+            ngx.say(data)
+        }
+    }
+--- request
+GET /t
+--- response_body
+l1_serializer returned an error: cannot transform
+nil
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: l1_serializer can be given as a :get() parameter
```

Comment:
An additional test here might be to assert it is a function again (since it is a different code path in the library)